relay-server/src/server.rs

```diff
@@ -1,143 +1,123 @@
-        let rm = self.istream().read_http_request()?;
-        let ostream = self.ostream();
-        let mut write = |text: &str| ostream.write(text.as_bytes());
-        let mut write_line = |line: &str| {
-            write(line)?;
-            write("\r\n")?;
-            Ok::<(), Error>(())
-        };
-        let mut write_response_line = || write_line("HTTP/1.1 200 OK");
-        match rm.method.as_str() {
+        let content = match request.method.as_str() {
             "GET" => {
-                let query = *rm.url.url_query().get("id").to_io_result("no id")?;
-                let msg = state
-                    .get(query.to_string())
-                    .map_or([].as_slice(), |v| v.as_slice());
-                let len = msg.len();
-                write_response_line()?;
-                write_line(format!("content-length:{len}").as_str())?;
-                write_line("")?;
-                ostream.write(msg)?;
+                let query = *request.url.url_query().get("id").to_io_result("no id")?;
+                self.0.get(query.to_string())
             }
             "POST" => {
-                state.post(rm.content);
-                write_response_line()?;
-                write_line("")?;
+                self.0.post(request.content);
+                Vec::default()
             }
             _ => return Err(Error::new(ErrorKind::InvalidData, "unknown HTTP method")),
         };
+        let response = Response::new(200, "OK".to_string(), Default::default(), content);
+        response.write(ostream)?;
+        ostream.flush()?;
         Ok(())
     }
+    // TODO: move this function to a `test` mod. 
+    pub fn call(&mut self, msg: &[u8]) -> Result<Vec<u8>, Error> {
+        let mut result = Vec::default();
+        let mut stream = msg.mem_io_stream(&mut result);
+        self.update(&mut stream)?;
+        if stream.i.position() != msg.len() as u64 {
+            return Err(Error::new(ErrorKind::InvalidData, "invalid request"));
+        }
+        Ok(result)
+    }
 }
 
-impl<T: Stream> ServerEx for T {}
-
 #[cfg(test)]
 mod test {
-    use crate::{server::ServerEx, state::State};
-    use std::{io::Cursor, str::from_utf8};
+    use std::str::from_utf8;
 
-    use super::Stream;
-
-    struct MockStream {
-        i: Cursor<&'static str>,
-        o: Cursor<Vec<u8>>,
-    }
-
-    trait MockStreamEx {
-        fn mock_stream(self) -> MockStream;
-    }
-
-    impl MockStreamEx for &'static str {
-        fn mock_stream(self) -> MockStream {
-            MockStream {
-                i: Cursor::new(self),
-                o: Default::default(),
-            }
-        }
-    }
-
-    impl Stream for MockStream {
-        type Read = Cursor<&'static str>;
-        type Write = Cursor<Vec<u8>>;
-        fn istream(&mut self) -> &mut Self::Read {
-            &mut self.i
-        }
-        fn ostream(&mut self) -> &mut Self::Write {
-            &mut self.o
-        }
-    }
+    use super::Server;
```

Comment:
Local test modules is the only place I advocate for using wildcard imports
```suggestion
    use super::*;
```